File: aoc2024/day04/star2.py
cnt = 0
data = [list(line) for line in open('day4.txt', 'r').read().splitlines()]
N = len(data)
for i in range(1, N - 1):
    for j in range(1, N - 1):
        if data[i][j] != 'A':
            continue
        x = data[i - 1][j - 1] + data[i + 1][j + 1]
        y = data[i - 1][j + 1] + data[i + 1][j - 1]
        if (x == 'MS' or x == 'SM') and (y == 'MS' or y == 'SM'):
            cnt += 1
print(cnt)

File: aoc2025/day4/star2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_set>

using namespace std;

vector<vector<char>> a;
int dir[8][2]{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}};
unordered_set<int> s;

int offset()
{
    return max(a.size(), a[0].size()) + 1;
}

void solve(int &n)
{
    bool removed{false};
    for (auto it = s.begin(); it != s.end();)
    {
        int r{*it / offset()}, c{*it % offset()}, cnt{0};
        for (const auto &[dr, dc] : dir)
        {
            if (s.count((r + dr) * offset() + c + dc))
            {
                ++cnt;
            }
        }
        if (cnt < 4)
        {
            it = s.erase(it);
            ++n;
            removed = true;
        }
        else
        {
            ++it;
        }
    }
    if (!removed)
    {
        return;
    }
    solve(n);
}

int main()
{
    ifstream in("day4.txt");
    string line;
    while (in >> line)
    {
        a.emplace_back(line.begin(), line.end());
    }
    in.close();
    s.reserve(a.size() * a[0].size());
    for (size_t r{0}; r < a.size(); ++r)
    {
        for (size_t c{0}, cnt{0}; c < a[r].size(); ++c, cnt = 0)
        {
            if (a[r][c] == '@')
            {
                s.insert(r * offset() + c);
            }
        }
    }
    int total{0};
    solve(total);
    cout << total << endl;
    return 0;
}
